opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -3890,6 +3942,94 @@ namespace Opm
+    {
+        const AutoICD& aicd = this->segmentSet()[seg].autoICD();
+
+        const int seg_upwind = this->upwinding_segments_[seg];
+        const std::vector<EvalWell>& phase_fractions = this->segment_phase_fractions_[seg_upwind];
+        const std::vector<EvalWell>& phase_viscosities = this->segment_phase_viscosities_[seg_upwind];
+        const std::vector<EvalWell>& phase_densities = this->segment_phase_densities_[seg_upwind];
+
+        EvalWell water_fraction = 0.;
+        EvalWell water_viscosity = 0.;
+        EvalWell water_density = 0.;
+        if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx)) {
+            const int water_pos = Indices::canonicalToActiveComponentIndex(FluidSystem::waterCompIdx);
+            water_fraction = phase_fractions[water_pos];
+            water_viscosity = phase_viscosities[water_pos];
+            water_density = phase_densities[water_pos];
+        }
+
+        EvalWell oil_fraction = 0.;
+        EvalWell oil_viscosity = 0.;
+        EvalWell oil_density = 0.;
+        if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx)) {
+            const int oil_pos = Indices::canonicalToActiveComponentIndex(FluidSystem::oilCompIdx);
+            oil_fraction = phase_fractions[oil_pos];
+            oil_viscosity = phase_viscosities[oil_pos];
+            oil_density = phase_densities[oil_pos];
+        }
+
+        EvalWell gas_fraction = 0.;
+        EvalWell gas_viscosity = 0.;
+        EvalWell gas_density = 0.;
+        if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
+            const int gas_pos = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
+            gas_fraction = phase_fractions[gas_pos];
+            gas_viscosity = phase_viscosities[gas_pos];
+            gas_density = phase_densities[gas_pos];
+        }
+
+        EvalWell density = segment_densities_[seg_upwind];
+        // WARNING
+        // We disregard the derivatives from the upwind density to make sure derivatives
+        // wrt. to different segments dont get mixed.
+        if (seg != seg_upwind) {
+            water_fraction.clearDerivatives();
+            water_viscosity.clearDerivatives();
+            water_density.clearDerivatives();
+            oil_fraction.clearDerivatives();
+            oil_viscosity.clearDerivatives();
+            oil_density.clearDerivatives();
+            gas_fraction.clearDerivatives();
+            gas_viscosity.clearDerivatives();
+            gas_density.clearDerivatives();
+            density.clearDerivatives();
+        }
+
+        using MathTool = MathToolbox<EvalWell>;
+        const EvalWell mixture_viscosity = MathTool::pow(water_fraction, aicd.waterViscExponent()) * water_viscosity
+                                         + MathTool::pow(oil_fraction, aicd.oilViscExponent()) * oil_viscosity
+                                         + MathTool::pow(gas_fraction, aicd.oilViscExponent()) * gas_viscosity;
```

Comment:
I think this should be gasViscExponent().

Reply:
good catch. Thanks.